File: searching/binary_search/binary_search.py
from typing import List


def search(arr: List, target: int, b: int = 0, e: int = None) -> int:

    e = len(arr)-1 if not e and e != 0 else e

    if b <= e:
        m = (b + e) // 2
        if arr[m] == target:
            return m
        elif arr[m] < target:
            return search(arr, target, m+1, e)
        else:
            return search(arr, target, b, m-1)

    return None

File: searching/binary_search/binary_search.cpp
#include <vector>
using namespace std; 

int search(vector<int> arr, int target, int b = 0, int e = -1){

    e = (e == -1) ? arr.size() : e; 

    if (b <= e){
        
        int m = b + e; 
        if (arr[m] == target){
            return m; 
        }
        else if (arr[m] < target){
            return search(arr, target, m+1, e);
        }
        else{
            return search(arr, target, b, m-1); 
        }
    }
    return -1;    
}
